src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -0,0 +1,805 @@
+        }
+
+        $objects = map($objects, function (string $key) use ($sourceBucket) {
+            return  self::formatAsS3URI($sourceBucket, $key);
+        });
+
+        $getObjectRequestCallback = null;
+        if (isset($config['get_object_request_callback'])) {
+            if (!is_callable($config['get_object_request_callback'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `get_object_request_callback` must be callable."
+                );
+            }
+
+            $getObjectRequestCallback = $config['get_object_request_callback'];
+        }
+
+        $failurePolicyCallback = null;
+        if (isset($config['failure_policy'])) {
+            if (!is_callable($config['failure_policy'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `failure_policy` must be callable."
+                );
+            }
+            
+            $failurePolicyCallback = $config['failure_policy'];
+        }
+
+        $promises = [];
+        $objectsDownloaded = 0;
+        $objectsFailed = 0;
+        $s3Delimiter = $config['s3_delimiter'] ?? '/';
+        foreach ($objects as $object) {
+            $bucketAndKeyArray = self::s3UriAsBucketAndKey($object);
+            $objectKey = $bucketAndKeyArray['Key'];
+            if ($s3Prefix !== null && str_contains($objectKey, $s3Delimiter)) {
+                if (!str_ends_with($s3Prefix, $s3Delimiter)) {
+                    $s3Prefix = $s3Prefix.$s3Delimiter;
+                }
+
+                $objectKey = substr($objectKey, strlen($s3Prefix));
+            }
+
+            // CONVERT THE KEY DIR SEPARATOR TO OS BASED DIR SEPARATOR
+            if (DIRECTORY_SEPARATOR !== $s3Delimiter) {
+                $objectKey = str_replace(
+                    $s3Delimiter,
+                    DIRECTORY_SEPARATOR,
+                    $objectKey
+                );
+            }
+
+            $destinationFile = $destinationDirectory . DIRECTORY_SEPARATOR . $objectKey;
+            if ($this->resolvesOutsideTargetDirectory($destinationFile, $objectKey)) {
+                throw new S3TransferException(
+                    "Cannot download key $objectKey "
+                    ."its relative path resolves outside the parent directory."
+                );
+            }
+
+            $requestArgs = $downloadDirectoryRequest->getGetObjectRequestArgs();
+            foreach ($bucketAndKeyArray as $key => $value) {
+                $requestArgs[$key] = $value;
+            }
+            if ($getObjectRequestCallback !== null) {
+                call_user_func($getObjectRequestCallback, $requestArgs);
+            }
+
+            $promises[] = $this->downloadFile(
+                new DownloadFileRequest(
+                    $destinationFile,
+                    $config['fails_when_destination_exists'] ?? false,
+                    new DownloadRequest(
+                        null,
```

Comment:
this might be more readable if you use named params or something like `fromArray` used for transforming the upload args for `uploadDirectory`